src/components/checkbox/examples/checkbox.tsx

```diff
@@ -45,9 +49,18 @@ export class CheckboxExample {
                             label="Toggle checked"
                             onChange={this.toggleChecked}
                         />
+                        <limel-checkbox
+                            checked={this.indeterminate}
+                            label="Toggle indeterminate"
```

Comment:
Oh, I guess the other checkboxes are also called "toggle…". I guess these suggestions are better applied when a little boyscouting has been done 😄

Anyway:

```suggestion
                            label="Indeterminate"
```

(I think the names comes from us originally using buttons, rather than checkboxes. For checkboxes, it makes more sense to skip the "toggle" bit.)

Reply:
Does it make sense to you if I also change toggleChecked to setChecked and remove the word toggle?

In my head, the tristate (not checked, checked, indeterminate) can be represented in many ways. 
Three ways are: 
1. enum 
2. a nullable boolean (null = indeterminate) 
3. using two boolean values (like MDC)

To me, indeterminate is part of the checked state and not part of other features such as disabled/required. Setting indeterminate will make the component to ignore the checked value altogether and I think we should use the same verb here (toggle, set or nothing) for both these properties.

Reply:
I'm having the exact same thoughts and then when I found out that MDC already got support using the "two boolean" variant I was really happy because adding it doesn't break anything.

I think the props and events can co-exist as separate, even though the visual effect of setting `indeterminate` will hide the visual effect of the state of `checked`. I didn't see the need of it at this moment, but an event for indeterminate could be added later on.

Consider this test case:

1. User clicks checkbox, sets `checked`, emits `changed` => `true`, visually shows `[v]`
2. User clicks something that causes the consumer code to set `indeterminate`, emits `indeterminateChanged` => `true`, visually shows `[-]`
3. User clicks checkbox, unsets `checked` and `indeterminate`, emits `changed` => `false` and `indeterminateChanged` => `false`, visually shows `[ ]`

Given this test case I need to make it untick the checkbox when clicking it. Now it remains in indeterminate state, which I find unexpected.

So, shortly, when interacting (keyboard spacebar/mouse click) with a checkbox with indeterminate state twice will make it transition: indeterminate -> not checked -> checked.
Please note that it only applies for user interaction. Modifying the `checked` prop `true` => `false` programmatically is not the same. 

The correct user interaction event handler would therefore ought to be: 

* with indeterminate set: unset checked and indeterminate
* with indeterminate unset: toggle checked

Reply:
PS. I don't mean to add an event for indeterminateChanged now. Only implement "unset indeterminate on user interaction".

Reply:
Also, what regards the events, while the event detail of the `change` event represents the current `checked` state, it will still be emitted for any change (disabled, required, indeterminate).

So without adding a separate event for indeterminate (we don't have any for disabled/required), I need to update the checkbox example indeterminate checkbox from that event handler to keep it in sync.
This have not been required for disabled/required because they can't be modified by user interaction while indeterminate can change from `true` => `false`, assuming the rules I stated above are correct.

I still think the `change` event is enough, it'll just make the example code a bit wonky not having a separate event + I need to make the indeterminate prop mutable.

Reply:
> I'm having the exact same thoughts and then when I found out that MDC already got support using the "two boolean" variant I was really happy because adding it doesn't break anything.

👍

> I think the props and events can co-exist as separate, even though the visual effect of setting `indeterminate` will hide the visual effect of the state of `checked`.

Yeah, I see no problem with that. We have the same thing where we have several components that can be either disabled or readonly. In many cases, a readonly component is _also_ disabled, just styled differently, so whether or not you explicitly set `disabled` doesn't matter for the appearance of the component.

> I didn't see the need of it at this moment, but an event for indeterminate could be added later on.
>
> Consider this test case:
> 
> 1. User clicks checkbox, sets `checked`, emits `changed` => `true`, visually shows `[v]`
> 2. User clicks something that causes the consumer code to set `indeterminate`, emits `indeterminateChanged` => `true`, visually shows `[-]`
> 3. User clicks checkbox, unsets `checked` and `indeterminate`, emits `changed` => `false` and `indeterminateChanged` => `false`, visually shows `[ ]`

I know this isn't your main point here, but I still want to explain why the latter half of step two would never happen:

The component should never emit events for state changes _initiated by the consumer_. If the user clicks the unchecked checkbox, we need to emit an event to tell the consumer that the state should change (probably, that's up to the consumer, really). But if the consumer sets `checked = true` programatically, the consumer already knows the new state, and so we don't emit an event about that. Since there's nothing the user can do to a checkbox, to set _that same_ checkbox to `indeterminate`, the case where we need to emit an event never appears.

Sure, if we perform DOM-manipulation, and argument could be made for emitting an event, but honestly, if DOM-manipulation is performed for any legitimate reason, I would expect that it's the consumer that's performing it, which means the consumer already knows about the new state, which in turn means we shouldn't emit an event for it.
 
> Given this test case I need to make it untick the checkbox when clicking it. Now it remains in indeterminate state, which I find unexpected.

Yeah, no, we definitely need to set indeterminate to false when the checkbox is clicked 👍
 
> So, shortly, when interacting (keyboard spacebar/mouse click) with a checkbox with indeterminate state twice will make it transition: indeterminate -> not checked -> checked.

I agree 👍

> Please note that it only applies for user interaction. Modifying the `checked` prop `true` => `false` programmatically is not the same.

Yeah, it's probably best to only change the value of the property actually changed when changed programatically, and not change the value of `indeterminate` just because `checked` was changed.

> The correct user interaction event handler would therefore ought to be:
> 
> * with indeterminate set: unset checked and indeterminate
> * with indeterminate unset: toggle checked

Or to be more verbose:

* checked: true, indeterminate: true -> checked: false, indeterminate: false - emit change event with value `false`
* checked: false, indeterminate: true -> checked: false, indeterminate: false - emit change event with value `false`
* checked: true, indeterminate: false -> checked: false, indeterminate: false - emit change event with value `false`
* checked: false, indeterminate: false -> checked: true, indeterminate: false - emit change event with value `true`

👍

Reply:
> Also, what regards the events, while the event detail of the `change` event represents the current `checked` state, it will still be emitted for any change (disabled, required, indeterminate).
> 
> So without adding a separate event for indeterminate (we don't have any for disabled/required), I need to update the checkbox example indeterminate checkbox from that event handler to keep it in sync. This have not been required for disabled/required because they can't be modified by user interaction while indeterminate can change from `true` => `false`, assuming the rules I stated above are correct.
> 
> I still think the `change` event is enough, it'll just make the example code a bit wonky not having a separate event + I need to make the indeterminate prop mutable.

The change event should not be emitted when the checkbox becomes `indeterminate`. I think my last answer probably explains why, and why the indeterminate prop shouldn't have to be mutable either (I think, again, _possibly_ with the exception for supporting DOM-manipulation, but let's cross that bridge when we determine we really do need to). If you disagree when you've read it, please comment about it again, and we'll discuss from there 👍

Reply:
How about required/disabled? (we emit the change event when those are modified even though the user can't interact with those props)

Reply:
🤨

I think (or hope) that you're confusing the change-handlers in the example controls for the events actually sent by limel-checkbox. limel-checkbox only send one type of events, according to the docs, and that's `change`.

The example component has checkboxes for triggering programmatic changes of the state. They have event handlers called things like `setRequired` and `setDisabled`, but they listen to the `change` event from their respective checkbox, neither of which are the checkbox "being demoed". It becomes a little confusing in this case, because both the component we are demoing and the components we use to control the demo, are limel-checkbox'es 😬

Reply:
Oh, yeah! Thanks for the heads up. I was totally having my breakpoint in the onChange handler of the checkbox class and not the change handler in the example file. I thought I was going crazy there for a moment. I got the change events for the checkboxes used in the example.

Reply:
After some further research I came to the conclusion that the MDC checkbox:
1. always unsets indeterminate on interaction
2. always toggles checked on interaction

Now the code is free from mutable props etc and the change event is emitted as expected. I also added an event printer to the example to make it super clear how the change event works with checked+indeterminate.